### ISSUE TITLE
Colour artefact with new Local Contrast tool

### ISSUE THREAD

**Issue by RawConvert** [Beep6581/RawTherapee#4504]
Using RawTherapee 5.4-79 it seems a colour glitch can occur when the Local Contrast tool is turned on.  The attached screen grab shows a red area at the top of the arch of the bridge.  This is not present until the tool is switched on with default settings.  Other areas are affected too, I rock near the centre of the image gets a blue area when the tool is turned on.  Increasing the "Amount" causes the colouration to spread and increase.  The raw and PP3 are here (too big to upload):-
https://filebin.net/9yy8mkfncgc7t6ld

![red-bit](https://user-images.githubusercontent.com/16868433/38676303-80186578-3e52-11e8-82a3-b18f79b4a193.png)

**Comment by agriggio** [comment]
I can't reproduce with the latest `dev`. Can you upload more info please? OS, `AboutThisBuild.txt` and your `options` file? Thanks!

**Comment by agriggio** [comment]
(sorry, my bad -- something is going on somewhere...)

**Comment by Beep6581** [comment]
Minimal reproducible PP3, can reproduce in any photo:
https://filebin.net/ywycsuywi90cgo6e/lc_artifacts.pp3

Related to the Lab CL curve setting high chroma on blacks:
![screenshot_20180412_145200](https://user-images.githubusercontent.com/5844619/38678386-2810c672-3e61-11e8-9478-0767fa6ae455.png)

**Comment by agriggio** [comment]
@Beep6581 thanks! I'll take a look ASAP. I'm not worried much by the artifact -- the curve is weird after all -- but this causes a segfault here, and this is bad.

**Comment by RawConvert** [comment]
Why is it weird?! - I'm making the green moss stuff greener and the mud browner!

**Comment by agriggio** [comment]
It's weird that you want to increas the chromaticity of black... I'd use something like this:

![curve](https://user-images.githubusercontent.com/11982086/38680278-33135b7a-3e66-11e8-8bbe-771bbb988c5a.jpeg)

But your curve is definitely not invalid -- it should not make RT crash!

**Comment by agriggio** [comment]
I've fixed the crash in f8c06863fe97759b6681f2c0589998227db590bd. The artifacts remain, but for that it's enough to use a "less weird" (please note the quotes :-) curve...

**Comment by RawConvert** [comment]
hi, well I'm not sure the curve with the very steep rise is that much less "weird" then mine.  However I've had a play with this and I still think there's something odd going on.  As per screen grab, I put a sample point where it goes red and watched what happened when L.C. is turned on.
Firstly, with your curve: This cures the red problem, so you're kind of right!  Turning LC on takes the sample from (9, .8, 0) to (10.6, .4, 0).  These are R,G,B %'s.
Then with my curve.  The sample starts at the same (9, .8, 0) interestingly. LC now takes the sample to (30.2, 0, 0) !
Why should LC do such different things to this bit of the image when the starting point is the same?
To me, LC is about making darker parts darker and lighter parts lighter, where they adjoin.  It seems odd the algorithm picks on this area and appreciably brightens it even though it's next to the sky!
![zeroed-curve](https://user-images.githubusercontent.com/16868433/38728762-af65078a-3f08-11e8-962c-fb77be715d91.png)
![my-curve](https://user-images.githubusercontent.com/16868433/38728775-b6b2a1e6-3f08-11e8-8a8b-af09b7422a4e.png)

**Comment by RawConvert** [comment]
I was just thinking, if I'd done the Local Contrast first, then increased the chromaticity in the dark areas, my question might be "why has the Lab CL curve done this strange thing?"  rather than "point the finger" at Local Contrast...  :)

**Comment by agriggio** [comment]
Hi,

> why has the Lab CL curve done this strange thing?

Maybe @Desmis can jelp?

**Comment by heckflosse** [comment]
The Lab CL curves changes Chromaticity C in dependance of Lightness L
Moving the left curve point upwards means the Chromaticity of the dark points will be increased.
That's perfectly fine. 
The question is: Why does the Lightness L increases as well?

**Comment by Desmis** [comment]
All the curves in Labadjustements have been focus some years ago (at least 4 or more) by exchanges with users, especially with Michael. After numerous exchanges we have reached these settings, but I do not see any problem in changing them if necessary. But for what ?

Jacques

**Comment by Desmis** [comment]
@heckflosse 

Ingo
I don't say it is normal that L change, but I think so :)

The values display take into account conversion Lab RGB at the end, if output profile is small, calculation  are done by the profile, and can increase L
If you check "avoid color shift", this effect is reduce
Jacques

**Comment by RawConvert** [comment]
Some more on Local Contrast.  Please load the jpeg below into RT with neutral profile and note all the red dots have the same colour and L value. Likewise green and blue. Now apply local contrast at say 100 radius and .3 amount.  Note that it alters the L value for the red and blue dots across the gradient.  BUT the L for the green dots hardly changes.  Surely this isn't right?
![dots](https://user-images.githubusercontent.com/16868433/38751988-a36bad32-3f51-11e8-959f-cf32bd9d4002.jpg)

**Comment by RawConvert** [comment]
Taking the dots jpeg and applying my CL curve, there are some artefacts top left -
![cl curve and dots](https://user-images.githubusercontent.com/16868433/38752702-1171c328-3f54-11e8-9938-1c2423e9aedc.png)
However to muddy the waters, I suspect that Gimp has introduced noise.  I used Gimp to create the gradient and draw the dots.  When I drew the gradient I left a small border left and right.  I have a feeling there is a little noise in the borders, no idea why.  However the artefacts above extend inwards beyond the border.
Now switch on Local Contrast (and keep CL curve) to get this -
![cl curve and local contrast](https://user-images.githubusercontent.com/16868433/38752885-94b4381a-3f54-11e8-9c7a-36b4f7cd85f1.png)
It seems a bit weird even taking the border issue into account!  (You can see the border I left)

**Comment by RawConvert** [comment]
@agriggio , @Beep6581 , @heckflosse , @Desmis , hi, do you have any thoughts on my comments above, perhaps I should have tagged ppl when I posted.

**Comment by heckflosse** [comment]
@RawConvert 

> Some more on Local Contrast. Please load the jpeg below into RT with neutral profile and note all the red dots have the same colour and L value. Likewise green and blue.

But the L value of the red dots is not the same as the L value of the green or blue dot.

For your example the calculation is roughly
```
Lout[x][y] = Lin[x][y] +  (Lin[x][y] - gaussianblur(Lin[x][y]) * amount 
```
Furthermore the vertical distance between the dots is < the radius (100) you set. That means for example the gaussian blurred L value of a point inside a green dot is influenced by the L values of the blue and red dot below and above and not just by the background.

**Comment by RawConvert** [comment]
Thanks @heckflosse , I darkened the green spots and played with the radius and I see what you mean.  So that puts to bed the stuff about the dots.
However I just noticed something else... not trying to be a nuisance here, honestly,  I cropped my test image to remove the dots and noticed a sudden change in the histogram going from radius 69 to 70 -
![hist-r69](https://user-images.githubusercontent.com/16868433/38881679-4663c1ee-4260-11e8-87f3-19b926e28822.png)
![hist-r70](https://user-images.githubusercontent.com/16868433/38881689-4e10a650-4260-11e8-9545-c30e33af5261.png)

Here is the file used (was a tif, saved as jpeg for uploading) -
![dots-crop](https://user-images.githubusercontent.com/16868433/38881912-dbbf8692-4260-11e8-9e2a-fc0e2b6c0faa.jpg)

Actually the histogram is jumping around all over the place as radius changes, maybe this is irrelevant!  I don't want to waste your time.

**Comment by heckflosse** [comment]
@RawConvert You don't waste my time. The sudden change for radius 70 is because we use double precision instead of single precision floating point for gauss from that radius upwards.
https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/gauss.cc#L1146

Edit: Maybe we could lower that threshold (70) a bit...

**Comment by RawConvert** [comment]
Would it be a significant overhead to use double for all radius values?

**Comment by heckflosse** [comment]
@RawConvert 
> Would it be a significant overhead to use double for all radius values?

Yes, it would. Double precision gauss is not vectorized. And even, if it would be vextorized, it would be still 2 times slower than single precision gauss.
